File: Stack+Queue/145. Binary Tree Postorder Traversal/145. Binary Tree Postorder Traversal.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def postorderTraversal(self, root):
        """
        :type root: TreeNode
        :rtype: List[int]
        """
        # 1. 处理特殊情况，当根节点为空时，直接返回空列表作为遍历结果
        res=[]
        
        if not root:
            return res
        
        # 2. 从根节点开始一路向左，将所有的途经结点均加入栈中，注意最后再向栈中加入一个空结点
        sta=[]
        
        pos=root
        while pos:
            sta.append(pos)
            pos=pos.left
        sta.append(pos)

        # 3. 从栈顶出栈一个结点temp，然后判断出栈后的新栈顶结点的右子结点right是否与temp相等，若相等则栈顶结点的右子树已经输出完成，这时无需做任何操作；若不相等，则这时栈顶结点的右子树尚未输出，这时从结点right开始一路向左，将途经的结点加入栈中，最后再向栈内加入一个空结点
        while sta:

            temp=sta[-1]
            sta.pop()

            if temp:
                res.append(temp.val)

            if sta:
                right=sta[-1].right

                if temp!=right:
                    while right:
                        sta.append(right)
                        right=right.left
                    sta.append(right)
            
        return res

File: Stack+Queue/145. Binary Tree Postorder Traversal/145. Binary Tree Postorder Traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */

// 二叉树的使用栈的前序，中序以及后序遍历是务必要熟练掌握的内容
// 基本算法原理: 
// 0. 创建一个栈，首先沿着根节点一路向左，将途经的结点全部入栈直到二叉树最左侧的一个结点为止，然后循环执行下列操作直到栈为空为止：
// 1. 从栈顶出栈一个结点，将该结点输出
// 2. 判断出栈后的栈顶结点的右子结点是否与步骤1中出栈的结点相等，若相等，则无需将该右子结点加入栈中，否则，将该右子结点加入栈中，并从该右子结点开始一路向左，将沿途的结点全部加入栈中

class Solution {
public:
    vector<int> postorderTraversal(TreeNode* root) {
        stack<TreeNode *> sta;
        vector<int> res;

        TreeNode *pos=root;

        while(pos)
        {
            sta.push(pos);
            pos=pos->left;
        }

        sta.push(pos);      // 注意后序遍历时需要将空结点也加入栈中

        while(sta.size())
        {
            TreeNode *temp=sta.top();

            if(temp)
            res.push_back(temp->val);

            sta.pop();
            
            if(sta.size())
            {
                TreeNode *right=sta.top()->right;  
                
                if(right!=temp)
                {
                    while(right)
                    {
                        sta.push(right);
                        right=right->left;
                    }
                    sta.push(right);
                }
            }
        }

        return res;
    }
};